### ISSUE TITLE
NEW: Intensity axes

### ISSUE THREAD

**Issue by vidartf** [hyperspy/hyperspy#546]
Currently there is no way (that I know of) to store a name or units for the "intensity axes" of data, e.g. "counts" for EDS data. In many cases this is not needed, but there is already some places in the code where axes titles are added manually to plots (`Signal.plot_explained_variance_ratio` and `Signal.plot_cumulative_explained_variance_ratio` sets it as "Explained variance ratio" and "Cumulative explained variance ratio" respectively). Another example where it would be useful could be when tracking positions, e.g.:

``` python
# Setup a mock signal with a moving gaussian with noise:
import hyperspy.hspy as hs
g = hs.components.Gaussian()
data = []
for i in xrange(2, 9):
    g.centre.value = float(i)
    data.append(g.function(np.linspace(0.0, 10.0)))
data = np.array(data)
data += 0.4*np.random.rand(*data.shape)
s = hs.signals.Spectrum(data)

# Setup energy axes to mimic spectroscopy data
s.axes_manager[1].scale = 10.0/50
s.axes_manager[1].unit = 'keV'
s.axes_manager[1].name = 'Energy'

# Track the position of the peak by using a gaussian component
m = hs.create_model(s)
m.append(g)
m.multifit()

# Finally plot the position as a function of the navigation axes
spos = g.centre.as_signal()
spos.plot()
# Here y-axis has the right values (position of peak along signal axes), but 
# units and name is lost, and mislabeled as 'Intensity'
```

I don't know how this would be best solved, but maybe a field on `AxesManager` would be appropriate. As the intensity axes doesn't need scale/size/offset (this is normally figured out automatically when plotting) I don't think we need a full `DataAxis`, but it might be worth considering.

**Comment by francisco-dlp** [comment]
What about simple adding it to the metadata? e.g.:

```python
metadata.Signal.quantity = "position"
metadata.Signal.units = "nm"

```

**Comment by vidartf** [comment]
Well, as most of the plotting code doesn't actually have access to the parent `Signal`, I figured `AxesManager` to be the intuitive place. Also, as far as I see it `AxesManager` is responsible for laying out the data into unit space. Further, for certain data, it is sometimes useful to switch the intensity and signal axes (i.e. go to f = x(y) instead of f = y(x)). If so, the intensity axes should be converted into a non-linear DataAxis (after sorting), and this would be a lot easier if it was already on `AxesManager`.

**Comment by francisco-dlp** [comment]
> Well, as most of the plotting code doesn't actually have access to the parent Signal, I figured AxesManager to be the intuitive place.

I agree, but we can workaround this technical issue.

> Also, as far as I see it AxesManager is responsible for laying out the data into unit space. 

I don't agree with this. HyperSpy deals with quantites that are mapped across dimensions e.g. particle counts mapped across energy loss, position, time. HyperSpy classifies those dimensions as "signal" and "navigation" dimensions because it is convenient to do so for visualization and analysis purposes, but it is always possible to reclassify those dimensions. The dimensions across which the quantity is mapped are stored in `axes_manager`. However, the quantity itself (e.g. the number of particles in this case) is not a dimension across which something is mapped. Also, most of the time is doesn't make sense to "reclassify" it as a "mapping dimension". For these reasons, I don't think that it has a place in `AxesManager`.

> Further, for certain data, it is sometimes useful to switch the intensity and signal axes (i.e. go to f = x(y) instead of f = y(x)). If so, the intensity axes should be converted into a non-linear DataAxis (after sorting), and this would be a lot easier if it was already on AxesManager.

If someone ever needs to exchnage the signal quantity with a mapping dimension, I don't see any issue with getting the quantity name and units from `metadata` instead of `AxesManager`.

**Comment by francisco-dlp** [comment]
Once we implement this, it'll be nice if we could use a quantities package to operate correctly on the units, what do you think? Any preferred package for this?

**Comment by vidartf** [comment]
> HyperSpy deals with quantites [...]

OK, I think I get the architecture. So, to plot data you need three pieces of information then: Quantities as a function of hyperdimensional coordinates (numpy array), a description of the HD space (`AxesManager`), and a description of the quantities that are mapped. All three need to be stored on `Signal` somewhere. Currently, for plotting, a data access function (typically `Signal.__call__()`) and an `AxesManager` is supplied. I think we should keep the hard separation between `Signal` and the plot classes as it is, which would then require some object to describe the quantity that can be passed around. Maybe just a simple object with a "name" attribute and a units object (units object preferably supplied by a mature package, which should also then go on `DataAxis`).

Browsing for unit packages, I see two alternative types: Ones that combine the numpy arrays and units, and ones that are units only. The advantage with first is that units will happen "automatically" for a lot of operations, also when working directly with the data. However, this might not always be wanted. If units are kept separate, then we need to decide whether we want to introduce such automatic units processing automatically for Signal operations (multiplication, division, integration, etc.). Then raw processing can still be performed directly on the data if wanted. If no such automatic processing is implemented, then users will have to manually process quantity units when appropriate. Either way, I think quantity names/descriptions need to be handled manually in all cases ("Intensity", "Position", "Time" "Velocity"). Which of these types do you think is the best solution?

**Comment by vidartf** [comment]
Note: As far as I can tell, `pint` is the package with the smartest strings processing/formatting which I'd assume would be vital for us. It seems it would support all the scenarios I pointed to in my last comment. However, it requires all units to be defined beforehand, and doesn't understand e.g. that "nm-1" is supposed to be `nm**-1`, so I think we might have to perform pre-parsing to some degree.

**Comment by thomasaarholt** [comment]
Did anything come of this? I just ran into a situation where it would be handy to change the label and units of `"Intensity"`.

**Comment by francisco-dlp** [comment]
There is no pull request to implement this yet. The implementation that I suggested above should be easy to implement, so I'll add the easy label to this proposal. Adding support for operating correctly in units using a quantities package could be left for another PR.

**Comment by ericpre** [comment]
I am wondering if it would be useful to be able to display intensities either in "counts" or "electrons" for example, just as in DM? if so, it may be necessary to add a scale attribute to `metadata.Signal`?

**Comment by francisco-dlp** [comment]
The scale attribute is already there: it is the [gain_factor](http://hyperspy.org/hyperspy-doc/current/user_guide/metadata_structure.html#variance-linear-model) in Variance_linear_model.

**Comment by ericpre** [comment]
Ok, thanks!

**Comment by francisco-dlp** [comment]
Fixed in #1240
